engine/src/main/java/org/pentaho/di/trans/steps/script/ScriptAddedFunctions.java

```diff
@@ -0,0 +1,2532 @@
+          }
+          sArg1 = (String) ArgList[0];
+          Format dfFormatter;
+          sArg2 = (String) ArgList[1];
+          sArg3 = (String) ArgList[2];
+          if ( sArg3.length() == 2 ) {
+            Locale dfLocale = EnvUtil.createLocale( sArg3 );
+            dfFormatter = new SimpleDateFormat( sArg2, dfLocale );
+            oRC = dfFormatter.parseObject( sArg1 );
+          } else {
+            throw new RuntimeException( "Locale " + sArg3 + " is not 2 characters long." );
+          }
+        } catch ( Exception e ) {
+          throw new RuntimeException( "Could not apply the local format for locale "
+            + sArg3 + " with the given format " + sArg2 + " on the string for " + sArg1 + " : " + e.getMessage() );
+        }
+        break;
+      case 4:
+        try {
+          if ( isNull( ArgList, new int[] { 0, 1, 2, 3 } ) ) {
+            return null;
+          } else if ( isUndefined( ArgList, new int[] { 0, 1, 2, 3 } ) ) {
+            return undefinedValue;
+          }
+          sArg1 = (String) ArgList[0];
+          DateFormat dfFormatter;
+          sArg2 = (String) ArgList[1];
+          sArg3 = (String) ArgList[2];
+          sArg4 = (String) ArgList[3];
+
+          // If the timezone is not recognized, java will automatically
+          // take GMT.
+          TimeZone tz = TimeZone.getTimeZone( sArg4 );
+
+          if ( sArg3.length() == 2 ) {
+            Locale dfLocale = EnvUtil.createLocale( sArg3 );
+            dfFormatter = new SimpleDateFormat( sArg2, dfLocale );
+            dfFormatter.setTimeZone( tz );
+            oRC = dfFormatter.parseObject( sArg1 );
+          } else {
+            throw new RuntimeException( "Locale " + sArg3 + " is not 2 characters long." );
+          }
+        } catch ( Exception e ) {
+          throw new RuntimeException( "Could not apply the local format for locale "
+            + sArg3 + " with the given format " + sArg2 + " on the string for " + sArg1 + " : " + e.getMessage() );
+        }
+        break;
+      default:
+        throw new RuntimeException( "The function call str2date requires 1, 2, 3, or 4 arguments." );
+    }
+    return oRC;
+  }
+
+  public static Object date2str( ScriptEngine actualContext, Bindings actualObject, Object[] ArgList,
+    Object FunctionContext ) {
+    Object oRC = new Object();
+    switch ( ArgList.length ) {
+      case 0:
+        throw new RuntimeException( "Please provide a valid date to the function call date2str." );
+      case 1:
+        try {
+          if ( isNull( ArgList ) ) {
+            return null;
+          } else if ( isUndefined( ArgList ) ) {
+            return undefinedValue;
+          }
+          java.util.Date dArg1 = (java.util.Date) ArgList[0];
+          if ( dArg1.equals( null ) ) {
+            return null;
+          }
+          Format dfFormatter = new SimpleDateFormat();
+          oRC = dfFormatter.format( dArg1 );
+        } catch ( Exception e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 1128, 1150, 1179.